pandas/core/algorithms.py

```diff
@@ -240,18 +245,25 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
         values = com._ensure_object(values)
         mask = com.isnull(values)
         keys, counts = htable.value_count_object(values, mask)
-        if not dropna:
+        if not dropna and mask.any():
             keys = np.insert(keys, 0, np.NaN)
             counts = np.insert(counts, 0, mask.sum())
 
     result = Series(counts, index=com._values_from_object(keys))
     if bins is not None:
-        # TODO: This next line should be more efficient
-        result = result.reindex(np.arange(len(cat.categories)), fill_value=0)
-        if not is_category:
-            result.index = bins[:-1]
+        if is_category:
+            ix = np.arange(len(cat.categories))
+            if not dropna and -1 in values:
```

Comment:
Is `value_counts` a good name for this new method of Categorical if all it does is fix the index after the value-counts series has already been built (in this case, by the line `result = Series(counts, index=com._values_from_object(keys))` above)? Or do you mean that most or all of the Categorical-related code in `algorithms.value_counts` should be moved to this new method?